lib/streamlit/runtime/pages_manager.py

```diff
@@ -0,0 +1,148 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import threading
+from pathlib import Path
+from typing import Callable, Final
+
+from blinker import Signal
+
+from streamlit.logger import get_logger
+from streamlit.source_util import PageInfo, get_pages
+from streamlit.util import calc_md5
+from streamlit.watcher import watch_dir
+
+_LOGGER: Final = get_logger(__name__)
+
+
+class PagesManagerV1:
+    is_watching_pages_dir: bool = False
+
+    # This is a static method because we only want to watch the pages directory
+    # once on initial load.
+    @staticmethod
+    def watch_pages_dir(pages_manager: PagesManager):
+        lock = threading.Lock()
+        with lock:
+            if PagesManagerV1.is_watching_pages_dir:
+                return
+
+            def _on_pages_changed(_path: str) -> None:
+                pages_manager.invalidate_pages_cache()
+
+            main_script_path = Path(pages_manager.main_script_path)
+            pages_dir = main_script_path.parent / "pages"
+            watch_dir(
+                str(pages_dir),
+                _on_pages_changed,
+                glob_pattern="*.py",
+                allow_nonexistent=True,
+            )
+            PagesManagerV1.is_watching_pages_dir = True
+
+
+class PagesManager:
+    def __init__(self, main_script_path, **kwargs):
+        self._cached_pages: dict[str, PageInfo] | None = None
```

Comment:
nit: it might be a good idea to define a type-alias for the key and give it a proper name, such as 
`type PageName = str` if the key is the page name, or `PageHash` or whatever and use that type for the key everywhere where we expect it. Then its easier to understand what we put in there semantically

Reply:
Love the type alias, and I'll apply it. I'm going to keep the dict for now cause the type is quite descriptive for it, and it makes sense (and it also aligns to the source_util phrasing).